ocaml/lambda/translcore.ml

```diff
@@ -1067,6 +1080,15 @@ and transl_apply ~scopes
           {ap with ap_args = ap.ap_args @ args; ap_loc = loc;
                    ap_region_close = pos; ap_mode = mode; ap_result_layout = result_layout }
     | lexp, _ ->
+      (* CR gyorsh: [assume_zero_alloc] is not used in the cases above but
+         Misplaced_attribute won't be reported for it.
+         Same for [@inlined] [@specialized] and tailcall.
+         It's fine for [Lsend] cases because [assume_zero_alloc] is
+         always false currently for them. I am not sure about Lapply case
+         above. Also not sure about inlined etc in the Lsend case.  *)
```

Comment:
I agree the `Lsend` cases aren't worth doing anything about.

The `Lapply` case above is optimizing `(f x) y` into `f x y` in some cases.

As an example, the current implementation checks this (which does not use the optimization) successfully:
```ocaml
let[@inline never] f x y = (x,y)

let[@zero_alloc] g () =
  let x = (f[@zero_alloc assume]) () () in
  x
```
But here are three alternate versions of the line beginning `let x = ...`, none of which check:

```ocaml
  let x = ((f[@zero_alloc assume]) ()) () in ...

  let x = ((f ())[@zero_alloc assume]) () in ...

  let x = (((f[@zero_alloc assume]) ())[@zero_alloc assume]) () in ...
```
Maybe all of these should work, in the case this optimization is applied?

Reply:
OK, yes, I can see the argument that this optimization does not necessarily preserve the meaning of the zero_alloc attribute and therefore it's safest to drop it (though I think maybe the last example is safe and could be supported in an ideal world).  But I am happy to put this problem off for another day.